internal/ingress/controller/nginx.go

```diff
@@ -424,11 +424,18 @@ Error: %v
 %v
 -------------------------------------------------------------------------------
 `, err, string(out))
+
+		// In case of error the temporal configuration file will
+		// be available up to five minutes after the error
+		go func(tmpFile string) {
```

Comment:
Can you limit the number of max possible goroutines so that we don’t potentially DOS ourselves.

Reply:
@ElvinEfendi actually, the goroutines will sleep 99% of the time, there is no gain in doing what you are proposing.

Reply:
Here's my worry:
Given a working ingress-nginx deployment in a cluster, someone makes a change that causes Nginx config test to fail. The controller in that case will log an error and spin up a goroutine to delete the temporary file. At the same time controller will retry generating new Nginx configuration. This will cause another goroutine to be span up. And eventually the controller will get OOMKilled (because we retry faster than 5 minutes the goroutines will accumulate). And this will break all the services.

> we need some queue aware of time

why can we not just check creation time of the file? Instead of spinning up a new goroutine here we can collect the filenames in a queue and at the start of controller we can start a "garbage collector goroutine" that periodically pops from the queue and checks if the file was created longer than 5 minutes ago and if so delete it.

Reply:
@ElvinEfendi ready for review

Reply:
> Here's my worry:
> Given a working ingress-nginx deployment in a cluster, someone makes a change that causes Nginx config test to fail. The controller in that case will log an error and spin up a goroutine to delete the temporary file. At the same time controller will retry generating new Nginx configuration. This will cause another goroutine to be span up. And eventually the controller will get OOMKilled (because we retry faster than 5 minutes the goroutines will accumulate). And this will break all the services.
> 
> > we need some queue aware of time
> 
> why can we not just check creation time of the file? Instead of spinning up a new goroutine here we can collect the filenames in a queue and at the start of controller we can start a "garbage collector goroutine" that periodically pops from the queue and checks if the file was created longer than 5 minutes ago and if so delete it.

@ElvinEfendi @aledbf I think I'm seeing this behavior in nginx-ingress [0].
it can be reproduced like this:
1. Create a service that listens to let's say 9500 and expose TCP port 9500
2. make the pod go to crashloopback (eg with a command sleep 2s)
3. the nginx-ingress-controller will go to status OOMKilled

Shall I create a new issue?

[0]
Release:    0.23.0
Build:      git-be1329b22)

Reply:
With these two manifests you can reproduce it:
nginx-ingress-manifest: https://gist.github.com/strigazi/2fd6a4f139b146f11d50f93811de490c
example-noop-server.yaml: https://gist.github.com/strigazi/d1d11f1efe91ce2045bf5d7517659924

Reply:
Since this a a reproducible issue for me, I'll open an new github issue so that other can find.